internal/elasticsearch/security/role_mapping.go

```diff
@@ -0,0 +1,240 @@
+package security
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients"
+	"github.com/elastic/terraform-provider-elasticstack/internal/models"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"strings"
+)
+
+func ResourceRoleMapping() *schema.Resource {
+	roleMappingSchema := map[string]*schema.Schema{
+		"id": {
+			Description: "Internal identifier of the resource",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"name": {
+			Description: "The name of the role mapping.",
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+		},
+		"enabled": {
+			Description: "Specifies whether the role mapping is enabled.",
+			Type:        schema.TypeBool,
+			Required:    true,
+		},
+		"roles": {
+			Description: " A list of role names that are granted to the users that match the role mapping rule. Default is [].",
+			Type:        schema.TypeSet,
+			Optional:    true,
+			Computed:    true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"role_templates": {
+			Description: "A list of mustache templates that will be evaluated to determine the roles names that should granted to the users that match the role mapping rules. Each record must be a valid JSON document.",
+			Type:        schema.TypeList,
+			Optional:    true,
+			MinItems:    1,
+			Elem: &schema.Schema{
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringIsJSON,
+			},
+		},
+		"rules": {
```

Comment:
Same goes about the rules, it might be better to provide the proper schema here for all the rules types which can be supported inside of the rules and in the worst case scenario just use json there.

Reply:
Not sure how this will work... Rules can be pretty complex, with types nested inside types. i.e
```
{
  "all": [
    {
      "field": {
        "username": "*"
      }
    },
    {
      "except": {
        "any": [
          {
            "field": {
              "username": "*"
            }
          },
          {
            "all": [
              {
                "field": {
                  "username": "*"
                }
              }
            ]
          }
        ]
      }
    },
    {
      "any": [
        {
          "field": {
            "username": "*"
          }
        }
      ]
    }
  ]
}
```

Because of my limited knowledge about TF providers, I'm not 100% sure how to write code that can validate this.

Reply:
Yeah, you're right. We still need to make some tradeoffs here. 
At least we should research this way if it's even possible and what it would take to make it work. If there is too much complexity we definitely will need to fall back to plain json.. 

Thanks a lot of working on this!